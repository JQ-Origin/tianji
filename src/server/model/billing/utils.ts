export interface CreditPackDefinition {
  id: string;
  name: string;
  variantId: string;
  credit: number;
  price: number;
  currency?: string;
}

export function parseCreditPacks(
  rawPacks: string | undefined
): CreditPackDefinition[] {
  if (!rawPacks) {
    return [];
  }

  try {
    const parsed = JSON.parse(rawPacks);
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed
      .map((item) => {
        const variantId = String(item.variantId ?? item.id ?? '');
        const credit = Number(item.credit ?? item.amount ?? 0);
        const price = Number(item.price ?? 0);

        if (!variantId || Number.isNaN(credit) || Number.isNaN(price)) {
          return null;
        }

        return {
          id: String(item.id ?? variantId),
          name: String(item.name ?? ''),
          variantId,
          credit,
          price,
          currency: item.currency ? String(item.currency) : undefined,
        };
      })
      .filter(Boolean) as CreditPackDefinition[];
  } catch (error) {
    console.error('Failed to parse LEMON_SQUEEZY_CREDIT_PACKS', error);
    return [];
  }
}
