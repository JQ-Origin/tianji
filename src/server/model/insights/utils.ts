import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { get, uniq } from 'lodash-es';
import { getDateArray } from '@tianji/shared';

/**
 * Process the grouped time series data
 * @param query - The query object
 * @param context - The context object
 * @param data - The database result
 * @returns The processed data
 */
export function processGroupedTimeSeriesData(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string },
  data: { date: string | null }[]
) {
  const { time, metrics, groups } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  let result: {
    name: string;
    [groupName: string]: any;
    data: {
      date: string;
      value: number;
    }[];
  }[] = [];

  for (const m of metrics) {
    if (groups.length > 0) {
      for (const g of groups) {
        const allGroupValue = uniq(
          data.map((item) => get(item, `%${g.value}`) as any)
        );

        result.push(
          ...allGroupValue.map((gv) => ({
            name: m.name,
            [g.value]: gv,
            data: getDateArray(
              data
                .filter((item) => get(item, `%${g.value}`) === gv)
                .map((item) => {
                  return {
                    value: Number(get(item, m.name)),
                    date: String(item.date),
                  };
                }),
              startAt,
              endAt,
              unit,
              timezone
            ),
          }))
        );
      }
    } else {
      result.push({
        name: m.name,
        data: getDateArray(
          data.map((item) => {
            return {
              value: Number(get(item, m.name)),
              date: String(item.date),
            };
          }),
          startAt,
          endAt,
          unit,
          timezone
        ),
      });
    }
  }

  return result;
}
