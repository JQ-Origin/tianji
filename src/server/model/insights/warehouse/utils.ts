import { z } from 'zod';
import { createPool, Pool } from 'mysql2/promise';
import { env } from '../../../utils/env.js';
import { Prisma } from '@prisma/client';

// MySQL date formats for different time units
export const MYSQL_DATE_FORMATS = {
  minute: '%Y-%m-%d %H:%i:00',
  hour: '%Y-%m-%d %H:00:00',
  day: '%Y-%m-%d',
  month: '%Y-%m-01',
  year: '%Y-01-01',
} as const;

export const dateTypeSchema = z.enum([
  'timestamp', // for example: 1739203200
  'timestampMs', // for example: 1739203200000
  'date', // for example: 2025-08-01
  'datetime', // for example: 2025-08-01 00:00:00
]);

const warehouseLongTableInsightsApplicationSchema = z.object({
  databaseUrl: z.string().optional(),
  name: z.string(),
  type: z.literal('longTable'), // long table
  eventTable: z.object({
    name: z.string(),
    eventNameField: z.string(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
  eventParametersTable: z.object({
    name: z.string(),
    eventNameField: z.string(),
    paramsNameField: z.string(),
    paramsValueField: z.string(),
    paramsValueNumberField: z.string().optional(),
    paramsValueStringField: z.string().optional(),
    paramsValueDateField: z.string().optional(),
    createdAtField: z.string(),
    createdAtFieldType: dateTypeSchema.default('timestampMs'),
    dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
  }),
});

const warehouseWideTableInsightsApplicationSchema = z.object({
  databaseUrl: z.string().optional(),
  name: z.string(),
  type: z.literal('wideTable'), // wide table
  tableName: z.string(),
  fields: z.array(
    z.object({
      name: z.string(),
      type: z.string().default('string'),
    })
  ),
  distinctField: z.string(),
  createdAtField: z.string(),
  createdAtFieldType: dateTypeSchema.default('timestampMs'),
  dateBasedCreatedAtField: z.string().optional(), // for improve performance, treat as date type
});

export const warehouseInsightsApplicationSchema = z.union([
  warehouseLongTableInsightsApplicationSchema,
  warehouseWideTableInsightsApplicationSchema,
]);

export type WarehouseInsightsApplication = z.infer<
  typeof warehouseInsightsApplicationSchema
>;

export type WarehouseLongTableInsightsApplication = z.infer<
  typeof warehouseLongTableInsightsApplicationSchema
>;

export type WarehouseWideTableInsightsApplication = z.infer<
  typeof warehouseWideTableInsightsApplicationSchema
>;

let applications: WarehouseInsightsApplication[] | null = null;
export function getWarehouseApplications() {
  if (!applications) {
    applications = warehouseInsightsApplicationSchema
      .array()
      .parse(JSON.parse(env.insights.warehouse.applicationsJson || '[]'));
  }

  return applications;
}

export function findWarehouseApplication(name: string) {
  return getWarehouseApplications().find((app) => app.name === name);
}

const connections = new Map<string, Pool>();
export function getWarehouseConnection(url = env.insights.warehouse.url) {
  if (!env.insights.warehouse.enable) {
    throw new Error('Warehouse is not enabled');
  }

  if (!url) {
    throw new Error('Warehouse url is not set');
  }

  let connection = connections.get(url);
  if (!connection) {
    connection = createPool(url);
    connections.set(url, connection);
  }

  return connection;
}

/**
 * Execute a data warehouse query
 * @param applicationId Application ID
 * @param sql Query SQL
 * @returns Query result
 */
export async function executeWarehouseQuery(
  applicationId: string,
  sql: Prisma.Sql
): Promise<any[]> {
  const application = findWarehouseApplication(applicationId);
  if (!application) {
    throw new Error(`Application ${applicationId} not found`);
  }

  const connection = getWarehouseConnection(application.databaseUrl);

  const [rows] = await connection.query(
    sql.sql.replaceAll('"', '`'), // avoid mysql and pg sql syntax error about double quote
    sql.values
  );

  return rows as any[];
}
