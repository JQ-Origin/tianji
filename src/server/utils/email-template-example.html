<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <meta http-equiv="X-UA-Compatible" content="ie=edge" />
  <title>Tianji Email Notification</title>
</head>
<body style="margin: 0; padding: 0; -webkit-text-size-adjust: none; text-size-adjust: none; width: 100%; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; -webkit-font-smoothing: antialiased; font-size: 16px; line-height: 1.6; color: #1a1a1a; background-color: #f3f4f6;">

  <table role="presentation" width="100%" cellpadding="0" cellspacing="0" border="0" style="padding: 40px 0;">
    <tr>
      <td align="center" valign="top">
        <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="600" style="max-width: 640px; margin: 0 auto; padding: 0; background-color: transparent;">
          <tr>
            <td align="center" style="padding: 0 0 24px 0;">
              <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="100%">
                <tr>
                  <td align="center">
                    <img src="https://tianji.msgbyte.com/img/logo@128.png" width="64" height="64" alt="Tianji" style="display: block; border: 0; height: auto; line-height: 100%; max-width: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; border-radius: 6px; margin-top: 16px;" />
                  </td>
                </tr>
                <tr>
                  <td align="center">
                    <h1 style="color: #0f172a; font-size: 28px; font-weight: 600; letter-spacing: -0.5px; margin: 0; padding: 0;">Tianji</h1>
                  </td>
                </tr>
                <tr>
                  <td align="center">
                    <p style="color: #64748b; font-size: 16px; margin-top: 4px; margin: 0; padding: 0;">Insight into everything</p>
                  </td>
                </tr>
              </table>
            </td>
          </tr>

          <tr>
            <td style="background-color: #ffffff; border-radius: 12px; padding: 48px; box-shadow: 0 4px 16px rgba(0, 0, 0, 0.08);">
              <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="100%">
                <tr>
                  <td>
                    <h2 style="color: #0f172a; font-size: 22px; font-weight: 600; margin: 0 0 24px 0; padding: 0; letter-spacing: -0.3px;">Hello, User!</h2>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p style="margin: 0 0 20px 0; padding: 0; color: #334155; line-height: 1.6;">
                      Thank you for using Tianji service. We are sending this email to notify you about important events in the system.
                    </p>
                  </td>
                </tr>

                <tr>
                  <td>
                    <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="100%" style="background-color: #ffffff; border-radius: 8px; margin: 20px 0; border-left: 4px solid #0f172a;">
                      <tr>
                        <td style="padding: 20px;">
                          <h3 style="margin: 0 0 12px 0; padding: 0; font-size: 18px; color: #0f172a; font-weight: 600;">System Monitoring Alert</h3>
                          <p style="margin: 0 0 12px 0; padding: 0; color: #334155;">Your application <span style="color: #0f172a; font-weight: 600;">Example Application</span> has experienced performance fluctuations in the past 5 minutes.</p>
                          <p style="margin: 0; padding: 0; font-size: 14px; color: #64748b;">Trigger time: 2023-11-10 15:30:45</p>
                        </td>
                      </tr>
                    </table>
                  </td>
                </tr>

                <tr>
                  <td>
                    <p style="margin: 0 0 20px 0; padding: 0; color: #334155; line-height: 1.6;">
                      You can click the button below to view detailed information and take appropriate actions.
                    </p>
                  </td>
                </tr>

                <tr>
                  <td align="center" style="padding: 20px 0;">
                    <table role="presentation" cellpadding="0" cellspacing="0" border="0">
                      <tr>
                        <td align="center" bgcolor="#0f172a" style="border-radius: 6px;">
                          <a href="https://tianji.msgbyte.com/dashboard" target="_blank" style="display: inline-block; font-weight: 500; text-align: center; background-color: #0f172a; color: #ffffff; text-decoration: none; padding: 12px 28px; border-radius: 6px; font-size: 16px;">View Details</a>
                        </td>
                      </tr>
                    </table>
                  </td>
                </tr>

                <tr>
                  <td>
                    <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="100%">
                      <tr>
                        <td style="height: 1px; background-color: #e5e7eb; margin: 24px 0;"></td>
                      </tr>
                    </table>
                  </td>
                </tr>

                <tr>
                  <td>
                    <p style="margin: 0 0 12px 0; padding: 0; color: #475569; font-size: 15px;">
                      If you did not expect to receive this email, please ignore it or contact the system administrator.
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p style="margin: 0; padding: 0; color: #475569; font-size: 15px;">
                      We hope you enjoy using the service,<br/>
                      Tianji Team
                    </p>
                  </td>
                </tr>
              </table>
            </td>
          </tr>

          <tr>
            <td style="padding: 32px 0;">
              <table role="presentation" cellpadding="0" cellspacing="0" border="0" width="100%">
                <tr>
                  <td align="center" style="padding-bottom: 16px;">
                    <a href="https://github.com/msgbyte/tianji" target="_blank" style="display: inline-block; margin: 0 6px;">
                      <img src="https://tianji.msgbyte.com/img/social/github.png" width="24" height="24" alt="GitHub" style="border: 0; height: auto; line-height: 100%; max-width: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; color: #0f172a;" />
                    </a>
                    <a href="https://twitter.com/moonrailgun" target="_blank" style="display: inline-block; margin: 0 6px;">
                      <img src="https://tianji.msgbyte.com/img/social/x.png" width="24" height="24" alt="Twitter" style="border: 0; height: auto; line-height: 100%; max-width: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; color: #0f172a;" />
                    </a>
                  </td>
                </tr>
                <tr>
                  <td align="center" style="padding-bottom: 16px;">
                    <a href="https://github.com/msgbyte/tianji" style="display: inline-block; font-size: 14px; color: #0f172a; text-decoration: none; margin: 0 8px;" target="_blank">GitHub</a>
                    <a href="https://tianji.msgbyte.com/docs/intro" style="display: inline-block; font-size: 14px; color: #0f172a; text-decoration: none; margin: 0 8px;" target="_blank">Documentation</a>
                  </td>
                </tr>
                <tr>
                  <td align="center">
                    <p style="font-size: 14px; color: #64748b; margin: 0; padding: 0;">
                      Â© 2023 Tianji. All rights reserved.
                    </p>
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>
