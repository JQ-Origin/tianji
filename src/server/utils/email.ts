import nodemailer from 'nodemailer';
import { logger } from './logger.js';
import { env } from './env.js';

export interface EmailOptions {
  to: string | string[];
  subject: string;
  text?: string;
  html?: string;
}

export async function sendEmail(options: EmailOptions) {
  try {
    const transporter = nodemailer.createTransport(env.smtp.server);

    const mailOptions = {
      from: env.smtp.from,
      to: options.to,
      subject: options.subject,
      text: options.text,
      html: options.html,
    };

    // TODO: create a email model to store email history

    const info = await transporter.sendMail(mailOptions);
    logger.info(`Email sent: ${info.messageId}`);
    return info;
  } catch (error) {
    logger.error('Failed to send email', error);
    throw new Error('Failed to send email');
  }
}
