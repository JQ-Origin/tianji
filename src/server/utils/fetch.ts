import * as http from 'http';
import * as https from 'https';
import * as url from 'url';

/**
 * Timing information for HTTP requests
 */
export interface RequestTimings {
  /** DNS lookup time in milliseconds */
  dns: number;
  /** TCP connection time in milliseconds */
  tcp: number;
  /** TLS handshake time in milliseconds */
  tls: number;
  /** Time to first byte (TTFB) in milliseconds */
  ttfb: number;
  /** Download time in milliseconds */
  download: number;
  /** Total request time in milliseconds */
  total: number;
}

/**
 * Response with timing information
 */
export interface TimedResponse<T = unknown> {
  /** Response data */
  data: T;
  /** HTTP status code */
  status: number;
  /** Response headers */
  headers: Record<string, string | string[]>;
  /** Detailed timing information */
  timings: RequestTimings;
  /** Request URL */
  url: string;
  /** HTTP method used */
  method: string;
}

/**
 * Options for timed fetch requests
 */
export interface TimedFetchOptions {
  /** Whether to parse JSON response automatically (default: true) */
  parseJson?: boolean;
  /** HTTP method */
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'HEAD' | 'OPTIONS';
  /** Request headers */
  headers?: Record<string, string>;
  /** Request timeout in milliseconds */
  timeout?: number;
  /** Request body as JSON */
  json?: Record<string, unknown>;
  /** Raw request body */
  body?: string | Buffer;
  /** Query parameters */
  searchParams?: Record<string, string | number | boolean>;
  /** Follow redirects (default: true) */
  followRedirect?: boolean;
  /** Maximum number of redirects (default: 10) */
  maxRedirects?: number;
  /** HTTP Agent for connection pooling and proxy support */
  httpsAgent?: https.Agent;
  /** Retry configuration */
  retry?: {
    limit?: number;
    methods?: string[];
    statusCodes?: number[];
  };
}

/**
 * Make an HTTP request with detailed timing information using native Node.js modules
 *
 * @param url - The URL to request
 * @param options - Request options
 * @returns Promise resolving to response with timing data
 *
 * @example
 * ```typescript
 * // Basic GET request
 * const result = await timedFetch('https://api.example.com/data');
 * console.log('Response:', result.data);
 * console.log('Total time:', result.timings.total, 'ms');
 *
 * // POST request with data
 * const postResult = await timedFetch('https://api.example.com/users', {
 *   method: 'POST',
 *   json: { name: 'John', email: 'john@example.com' }
 * });
 *
 * // Custom configuration with agent
 * const customAgent = new https.Agent({ keepAlive: true });
 * const customResult = await timedFetch('https://api.example.com/data', {
 *   method: 'GET',
 *   headers: { 'Authorization': 'Bearer token' },
 *   timeout: 5000,
 *   agent: customAgent,
 *   parseJson: true,
 *   retry: { limit: 3 }
 * });
 * ```
 */
export async function timedFetch<T = unknown>(
  requestUrl: string,
  options: TimedFetchOptions = {}
): Promise<TimedResponse<T>> {
  const {
    parseJson = true,
    method = 'GET',
    headers = {},
    timeout = 30000,
    json,
    body,
    searchParams,
    followRedirect = true,
    maxRedirects = 10,
    httpsAgent,
    retry = { limit: 0 },
  } = options;

  let attempts = 0;
  const maxAttempts = (retry.limit || 0) + 1;

  while (attempts < maxAttempts) {
    attempts++;

    try {
      return await makeRequest<T>(
        requestUrl,
        method,
        headers,
        timeout,
        json,
        body,
        searchParams,
        parseJson,
        followRedirect,
        maxRedirects,
        httpsAgent
      );
    } catch (error) {
      if (attempts >= maxAttempts) {
        throw error;
      }
      // Simple retry delay
      await new Promise((resolve) => setTimeout(resolve, 1000 * attempts));
    }
  }

  throw new Error('Max retry attempts reached');
}

async function makeRequest<T>(
  requestUrl: string,
  method: string,
  headers: Record<string, string>,
  timeout: number,
  json?: Record<string, unknown>,
  body?: string | Buffer,
  searchParams?: Record<string, string | number | boolean>,
  parseJson: boolean = true,
  followRedirect: boolean = true,
  maxRedirects: number = 10,
  httpsAgent?: https.Agent,
  redirectCount: number = 0
): Promise<TimedResponse<T>> {
  return new Promise((resolve, reject) => {
    // Parse URL
    let parsedUrl: url.URL;
    try {
      parsedUrl = new url.URL(requestUrl);
    } catch (error) {
      reject(new Error(`Invalid URL: ${requestUrl}`));
      return;
    }

    // Add search params to URL
    if (searchParams) {
      Object.entries(searchParams).forEach(([key, value]) => {
        parsedUrl.searchParams.set(key, String(value));
      });
    }

    const isHttps = parsedUrl.protocol === 'https:';
    const httpModule = isHttps ? https : http;

    // Prepare request body
    let requestBody: string | Buffer | undefined;
    const requestHeaders = { ...headers };

    if (json) {
      requestBody = JSON.stringify(json);
      requestHeaders['Content-Type'] = 'application/json';
      requestHeaders['Content-Length'] =
        Buffer.byteLength(requestBody).toString();
    } else if (body) {
      requestBody = body;
      if (typeof body === 'string') {
        requestHeaders['Content-Length'] = Buffer.byteLength(body).toString();
      } else {
        requestHeaders['Content-Length'] = body.length.toString();
      }
    }

    // Timing measurements
    const timings = {
      start: performance.now(),
      dnsLookup: 0,
      tcpConnect: 0,
      tlsConnect: 0,
      requestSent: 0,
      firstByte: 0,
      end: 0,
    };

    // Request options
    const requestOptions: http.RequestOptions = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port || (isHttps ? 443 : 80),
      path: parsedUrl.pathname + parsedUrl.search,
      method: method.toUpperCase(),
      headers: requestHeaders,
      timeout,
    };

    // Add agent if provided
    if (httpsAgent !== undefined && isHttps) {
      requestOptions.agent = httpsAgent;
    }

    const req = httpModule.request(requestOptions, (res) => {
      timings.firstByte = performance.now();

      // Handle redirects
      if (
        followRedirect &&
        res.statusCode &&
        res.statusCode >= 300 &&
        res.statusCode < 400 &&
        res.headers.location
      ) {
        if (redirectCount >= maxRedirects) {
          reject(new Error(`Too many redirects (${redirectCount})`));
          return;
        }

        const redirectUrl = url.resolve(requestUrl, res.headers.location);
        makeRequest<T>(
          redirectUrl,
          method,
          headers,
          timeout,
          json,
          body,
          searchParams,
          parseJson,
          followRedirect,
          maxRedirects,
          httpsAgent,
          redirectCount + 1
        )
          .then(resolve)
          .catch(reject);
        return;
      }

      const chunks: Buffer[] = [];

      res.on('data', (chunk: Buffer) => {
        chunks.push(chunk);
      });

      res.on('end', () => {
        timings.end = performance.now();

        const rawData = Buffer.concat(chunks);
        let responseData: T;

        try {
          if (
            parseJson &&
            res.headers['content-type']?.includes('application/json')
          ) {
            responseData = JSON.parse(rawData.toString('utf8')) as T;
          } else {
            responseData = rawData.toString('utf8') as T;
          }
        } catch (error) {
          responseData = rawData.toString('utf8') as T;
        }

        // Calculate timing differences
        const calculatedTimings: RequestTimings = {
          dns: Math.max(0, timings.dnsLookup - timings.start),
          tcp: Math.max(0, timings.tcpConnect - timings.dnsLookup),
          tls: isHttps
            ? Math.max(0, timings.tlsConnect - timings.tcpConnect)
            : 0,
          ttfb: Math.max(0, timings.firstByte - timings.start),
          download: Math.max(0, timings.end - timings.firstByte),
          total: Math.max(0, timings.end - timings.start),
        };

        // Convert headers to the expected format
        const responseHeaders: Record<string, string | string[]> = {};
        Object.entries(res.headers).forEach(([key, value]) => {
          if (value !== undefined) {
            responseHeaders[key] = value;
          }
        });

        const response: TimedResponse<T> = {
          data: responseData,
          status: res.statusCode || 0,
          headers: responseHeaders,
          timings: calculatedTimings,
          url: requestUrl,
          method: method.toUpperCase(),
        };

        resolve(response);
      });

      res.on('error', (error) => {
        reject(new Error(`Response error: ${error.message}`));
      });
    });

    // Set up timing event listeners
    req.on('lookup', () => {
      timings.dnsLookup = performance.now();
    });

    req.on('connect', () => {
      timings.tcpConnect = performance.now();
    });

    req.on('secureConnect', () => {
      timings.tlsConnect = performance.now();
    });

    req.on('finish', () => {
      timings.requestSent = performance.now();
    });

    req.on('timeout', () => {
      req.destroy();
      reject(new Error(`Request timeout after ${timeout}ms`));
    });

    req.on('error', (error) => {
      reject(new Error(`Request error: ${error.message}`));
    });

    // Write request body if present
    if (requestBody) {
      req.write(requestBody);
    }

    req.end();
  });
}
