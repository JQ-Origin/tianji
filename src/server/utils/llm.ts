import { Prisma } from '@prisma/client';

const contextWindows = await import('./model_prices_and_context_window.json', {
  with: { type: 'json' },
}).then((res) => res.default);

export function getLLMCostDecimal(
  model: string,
  inputToken: number,
  outputToken: number
): Prisma.Decimal {
  const input = new Prisma.Decimal(inputToken);
  const output = new Prisma.Decimal(outputToken);

  const contextWindow =
    contextWindows[model as keyof typeof contextWindows] ??
    contextWindows['gpt-4o-mini'];

  if (
    'input_cost_per_token' in contextWindow &&
    'output_cost_per_token' in contextWindow
  ) {
    return input
      .mul(contextWindow.input_cost_per_token)
      .add(output.mul(contextWindow.output_cost_per_token));
  } else {
    return new Prisma.Decimal(0);
  }
}
