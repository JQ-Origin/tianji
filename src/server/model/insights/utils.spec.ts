import { describe, expect, test } from 'vitest';
import { processGroupedTimeSeriesData } from './utils.js';

describe('processGroupedTimeSeriesData', () => {
  test('simple test case', () => {
    const query = {
      workspaceId: 'clnzoxcy10001vy2ohi4obbi0',
      insightId: 'cma0pdi2w6j5n1xy5y9qez5zc',
      insightType: 'website' as const,
      metrics: [{ name: '$all_event', math: 'sessions' as const }],
      filters: [],
      groups: [{ value: 'depth', type: 'number' as const }],
      time: {
        startAt: 1745769600000,
        endAt: 1746460799999,
        unit: 'day' as const,
      },
    };

    // database result
    const data = [
      { date: '2025-04-29', '%depth': 18.0, $all_event: 1 },
      { date: '2025-04-30', '%depth': 1.0, $all_event: 20 },
      { date: '2025-04-30', '%depth': 2.0, $all_event: 11 },
      { date: '2025-04-30', '%depth': 3.0, $all_event: 7 },
      { date: '2025-04-30', '%depth': 4.0, $all_event: 3 },
      { date: '2025-04-30', '%depth': 5.0, $all_event: 3 },
      { date: '2025-04-30', '%depth': 6.0, $all_event: 2 },
      { date: '2025-04-30', '%depth': 7.0, $all_event: 2 },
      { date: '2025-04-30', '%depth': 8.0, $all_event: 1 },
      { date: '2025-04-30', '%depth': 9.0, $all_event: 1 },
      { date: '2025-05-01', '%depth': 1.0, $all_event: 1 },
      { date: '2025-05-01', '%depth': 2.0, $all_event: 1 },
      { date: '2025-05-01', '%depth': 3.0, $all_event: 1 },
      { date: '2025-05-02', '%depth': 1.0, $all_event: 4 },
      { date: '2025-05-02', '%depth': 2.0, $all_event: 1 },
      { date: '2025-05-02', '%depth': 3.0, $all_event: 1 },
    ];

    const result = processGroupedTimeSeriesData(
      query,
      { timezone: 'utc' },
      data
    );

    expect(result).toMatchSnapshot();
  });
});
