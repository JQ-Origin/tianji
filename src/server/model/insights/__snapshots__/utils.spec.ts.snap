// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processGroupedTimeSeriesData > simple test case 1`] = `
[
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 18,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 20,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 4,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 1,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 11,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 2,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 7,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 3,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 3,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 4,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 3,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 5,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 2,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 6,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 2,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 7,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 8,
    "name": "$all_event",
  },
  {
    "data": [
      {
        "date": "2025-04-27 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-28 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-04-29 00:00:00",
        "value": 1,
      },
      {
        "date": "2025-04-30 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-01 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-02 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-03 00:00:00",
        "value": 0,
      },
      {
        "date": "2025-05-04 00:00:00",
        "value": 0,
      },
    ],
    "depth": 9,
    "name": "$all_event",
  },
]
`;
